### PR TITLE
fix remainder of failing integration tests

### PR DESCRIPTION
**Pull request by JacobOscarGunnarsson**
This PR fixes:

- checks for correct error messages in tests depending on platform
- moves invalid string indexing on results and lists to separate tests. They're also more elaborated on the expected behaviour
- adds a suggestion to change the `select` helper method
- updates readme for setting up react-native local tests

### CODE REVIEW

File: integration-tests/tests/src/utils/select.ts [hunk 1]
```diff
@@ -24,14 +24,14 @@ type PlatformValues<T> = Partial<Record<keyof KnownEnvironment, T>> & {
  * @param platformValues containing value for each targeted platform.
  * @returns the value for the platform the test is running on.
  */
```
Comment (JacobOscarGunnarsson) [inline integration-tests/tests/src/utils/select.ts:26]: this change may not be necessary, I realised when doing some testing that if `default == undefined` this helper function won't work. But then I'd like ism input on this part as having `any` as return type is not necessarily the best..

File: integration-tests/tests/src/tests/sync/sync-session.ts [hunk 2]
```diff
@@ -353,7 +354,9 @@ describe("SessionTest", () => {
     afterEach(() => Realm.clearTestState());
     it("can set custom logging function", async function (this: AppContext) {
       // setting a custom logging function must be done immediately after instantiating an app
-      const app = new Realm.App(this.app.id);
```
Comment (JacobOscarGunnarsson) [inline integration-tests/tests/src/tests/sync/sync-session.ts:356]: if I remember this failed in electron saying that the app could not be found

File: integration-tests/tests/src/tests/list.ts [hunk 3]
```diff
@@ -521,6 +522,24 @@ describe("Lists", () => {
...
+        const obj = this.realm.create<ILinkTypeSchema>("LinkTypesObject", {
+          objectCol: { doubleCol: 1 },
+          objectCol1: { doubleCol: 2 },
+          arrayCol: [{ doubleCol: 3 }],
+        });
+
+        //React native returns an empty object upon invalid indexing.
+        if (environment.reactNative) {
+          //@ts-expect-error TYPEBUG: indexing by string on results is not allowed typewise
+          expect(Object.keys(obj?.arrayCol[""]).length).equals(0);
```
Comment (kraenhansen) [inline integration-tests/tests/src/tests/list.ts:536]: In my opinion:
- the long term solution is to figure out why this is happening. This behavior shouldn't be platform dependent.
- the short term solution could be creating an issue to track this and disable the test for now, instead of encoding an unexpected behavior as a success.

File: integration-tests/tests/src/tests/sync/app.ts [hunk 4]
```diff
@@ -108,9 +108,10 @@ describe("App", () => {
       const credentials = Realm.Credentials.anonymous();
       await expect(app.logIn(credentials)).to.be.rejectedWith(
         select({
-          reactNative: "Network request failed: Could not connect to the server",
-          default:
+          reactNative: new RegExp("Network request failed.*connect"),
+          default: new RegExp(
```
Comment (kraenhansen) [inline integration-tests/tests/src/tests/sync/app.ts:112]: Why are there not simple strings? If they must be regexp consider using the `/ /` syntax.

File: integration-tests/tests/src/tests/sync/open-behavior.ts [hunk 5]
```diff
@@ -541,7 +541,8 @@ describe("OpenBehaviour", function () {
       await openPromise2;
       throw new Error("openPromise2 should have been rejected..");
     } catch (err: any) {
-      expect(err.message).equals("Operation canceled");
+      //platforms either return "Operation canceled" or "Operation Canceled"
+      expect((err.message as string).toLowerCase()).equals("operation canceled");
```
Comment (kraenhansen) [inline integration-tests/tests/src/tests/sync/open-behavior.ts:545]: Would it be easy to find the place this differs and solve it or is this thrown by the runtime?

File: integration-tests/tests/src/utils/select.ts [hunk 6]
```diff
@@ -24,14 +24,14 @@ type PlatformValues<T> = Partial<Record<keyof KnownEnvironment, T>> & {
...
-export function select<T>(platformValues: PlatformValues<T>): T {
+export function select<T>(platformValues: PlatformValues<T>): any {
```
Comment (kraenhansen) [inline integration-tests/tests/src/utils/select.ts:27]: Why this change? It'll be harder to determine a type mismatch in the calling code.

File: integration-tests/tests/src/utils/select.ts (hunk 1 above)
Comment (gagik) [reply integration-tests/tests/src/utils/select.ts:26]: What do you mean by `default == undefined`? I am unsure how this affects `select`'s type. If this is about the automatic derivation of the type from given values, you can manually specify i.e. `select<String | undefined>({default: "test"}})` so there is no need for `any` as a return type.

File: integration-tests/tests/src/tests/sync/open-behavior.ts (hunk 5 above)
Comment (JacobOscarGunnarsson) [reply integration-tests/tests/src/tests/sync/open-behavior.ts:545]: Can't find this anywhere in the project so my assumption is that it's thrown by the runtime, especially since the message varies

File: integration-tests/tests/src/tests/sync/app.ts (hunk 4 above)
Comment (JacobOscarGunnarsson) [reply integration-tests/tests/src/tests/sync/app.ts:112]: The reactNative one must be a regex. I skimmed through the documentation and thought that the `new RegExp` was the easiest readable declaration, but this is just based on my flavour so I can go with `/ /` as well

Comment (gagik) [reply integration-tests/tests/src/tests/sync/app.ts:112]: Why does it need to be regex? Could it not just be `"Network request failed"`?

Comment (JacobOscarGunnarsson) [reply integration-tests/tests/src/tests/sync/app.ts:112]: yeah fair, thought it was too generic at first, but in hindsight, my regex doesn't provide much more detailed information

**Review by takameyer — APPROVED** [review]
LGTM

File: integration-tests/tests/src/tests/sync/open-behavior.ts (hunk 5 above)
Comment (takameyer) [reply integration-tests/tests/src/tests/sync/open-behavior.ts:545]: I was able to find some generic error codes within the `boost` library in the `ios/Pods`, which contained an "Operation canceled" string.  Seems the message is out of our hands and is indeed platform specific.

File: integration-tests/tests/src/tests/list.ts (hunk 3 above)
Comment (kraenhansen) [reply integration-tests/tests/src/tests/list.ts]: @takameyer was this addressed?

Comment (kraenhansen) [reply integration-tests/tests/src/tests/list.ts]: Never mind, I updated my UI and found the change I needed to see 👍😅
